### ISSUE TITLE
BLAS/LAPACK routine 'DTRTI2' gave error code -1 in test-omega-chol.R:14

### ISSUE THREAD

**Issue by bruc** [nlmixrdevelopment/RxODE#84]
I'm getting a new error when running R CMD check on the current release of RxODE (version 0.8.0-9) (see below). I had run this check earlier this month (March 12) using RHEL 6 and the latest versions of devtoolset-8 from the Redhat Special collections.

When I checked the test runs on CRAN,  (see [Debian Devel gcc test run](https://www.r-project.org/nosvn/R.check/r-devel-linux-x86_64-debian-gcc/RxODE-00check.html) ), it shows the same errors that I'm getting. 

── 1. Error: omega chol (@test-omega-chol.R#14)  ───────────────────────────────
BLAS/LAPACK routine 'DTRTI2' gave error code -1
1: expect_equal(v$omega, mcov) at testthat/test-omega-chol.R:14
2: quasi_label(enquo(object), label)
3: eval_bare(get_expr(quo), get_env(quo))
4: v$omega
5: `$.rxSymInvCholEnv`(v, omega)

── 2. Error: omega chol (@test-omega-chol.R#14)  ───────────────────────────────
BLAS/LAPACK routine 'DTRTI2' gave error code -1
1: expect_equal(v$omega, mcov) at testthat/test-omega-chol.R:14
2: quasi_label(enquo(object), label)
3: eval_bare(get_expr(quo), get_env(quo))
4: v$omega
5: `$.rxSymInvCholEnv`(v, omega)

── 3. Error: omega chol (@test-omega-chol.R#14)  ───────────────────────────────
BLAS/LAPACK routine 'DTRTI2' gave error code -1
1: expect_equal(v$omega, mcov) at testthat/test-omega-chol.R:14
2: quasi_label(enquo(object), label)
3: eval_bare(get_expr(quo), get_env(quo))
4: v$omega
5: `$.rxSymInvCholEnv`(v, omega)

── 4. Error: omega chol (@test-omega-chol.R#14)  ───────────────────────────────
BLAS/LAPACK routine 'DTRTI2' gave error code -1
1: expect_equal(v$omega, mcov) at testthat/test-omega-chol.R:14
2: quasi_label(enquo(object), label)
3: eval_bare(get_expr(quo), get_env(quo))
4: v$omega
5: `$.rxSymInvCholEnv`(v, omega)

── 5. Error: omega chol (@test-omega-chol.R#14)  ───────────────────────────────
BLAS/LAPACK routine 'DTRTI2' gave error code -1
1: expect_equal(v$omega, mcov) at testthat/test-omega-chol.R:14
2: quasi_label(enquo(object), label)
3: eval_bare(get_expr(quo), get_env(quo))
4: v$omega
5: `$.rxSymInvCholEnv`(v, omega)

── 6. Error: omega chol (@test-omega-chol.R#14)  ───────────────────────────────
BLAS/LAPACK routine 'DTRTI2' gave error code -1
1: expect_equal(v$omega, mcov) at testthat/test-omega-chol.R:14
2: quasi_label(enquo(object), label)
3: eval_bare(get_expr(quo), get_env(quo))
4: v$omega
5: `$.rxSymInvCholEnv`(v, omega)

── 7. Error: omega chol (@test-omega-chol.R#14)  ───────────────────────────────
BLAS/LAPACK routine 'DTRTI2' gave error code -1
1: expect_equal(v$omega, mcov) at testthat/test-omega-chol.R:14
2: quasi_label(enquo(object), label)
3: eval_bare(get_expr(quo), get_env(quo))
4: v$omega
5: `$.rxSymInvCholEnv`(v, omega)

── 8. Error: omega chol (@test-omega-chol.R#14)  ───────────────────────────────
BLAS/LAPACK routine 'DTRTI2' gave error code -1
1: expect_equal(v$omega, mcov) at testthat/test-omega-chol.R:14
2: quasi_label(enquo(object), label)
3: eval_bare(get_expr(quo), get_env(quo))
4: v$omega
5: `$.rxSymInvCholEnv`(v, omega)

── 9. Error: omega chol (@test-omega-chol.R#14)  ───────────────────────────────
BLAS/LAPACK routine 'DTRTI2' gave error code -1
1: expect_equal(v$omega, mcov) at testthat/test-omega-chol.R:14
2: quasi_label(enquo(object), label)
3: eval_bare(get_expr(quo), get_env(quo))
4: v$omega
5: `$.rxSymInvCholEnv`(v, omega)

── 10. Error: omega chol (@test-omega-chol.R#14)  ──────────────────────────────
BLAS/LAPACK routine 'DTRTI2' gave error code -1
1: expect_equal(v$omega, mcov) at testthat/test-omega-chol.R:14
2: quasi_label(enquo(object), label)
3: eval_bare(get_expr(quo), get_env(quo))
4: v$omega
5: `$.rxSymInvCholEnv`(v, omega)

── 11. Error: omega chol (@test-omega-chol.R#14)  ──────────────────────────────
BLAS/LAPACK routine 'DTRTI2' gave error code -1
1: expect_equal(v$omega, mcov) at testthat/test-omega-chol.R:14
2: quasi_label(enquo(object), label)
3: eval_bare(get_expr(quo), get_env(quo))
4: v$omega
5: `$.rxSymInvCholEnv`(v, omega)

── 12. Error: omega chol (@test-omega-chol.R#14)  ──────────────────────────────
BLAS/LAPACK routine 'DTRTI2' gave error code -1
1: expect_equal(v$omega, mcov) at testthat/test-omega-chol.R:14
2: quasi_label(enquo(object), label)
3: eval_bare(get_expr(quo), get_env(quo))
4: v$omega
5: `$.rxSymInvCholEnv`(v, omega)

**Comment by mattfidler** [comment]
I think that blas or more likely RcppArmadillo was updated and doesn't work correctly.

Can you help me figure out where this problem occurs by posting the `sessionInfo()`?

**Comment by mattfidler** [comment]
This source code for the BLAS/Dtri2.f is here:

http://www.netlib.org/lapack/lapack-3.1.1/html/dtrti2.f.html;  It seems to be failing the upper triangular test:

```fortran
    INFO = 0
      UPPER = LSAME( UPLO, 'U' )
      NOUNIT = LSAME( DIAG, 'N' )
      IF( .NOT.UPPER .AND. .NOT.LSAME( UPLO, 'L' ) ) THEN
         INFO = -1
      ELSE IF( .NOT.NOUNIT .AND. .NOT.LSAME( DIAG, 'U' ) ) THEN
         INFO = -2
      ELSE IF( N.LT.0 ) THEN
         INFO = -3
      ELSE IF( LDA.LT.MAX( 1, N ) ) THEN
         INFO = -5
      END IF
      IF( INFO.NE.0 ) THEN
         CALL XERBLA( 'DTRTI2', -INFO )
         RETURN
      END IF
```

**Comment by bruc** [comment]
Here's the sessionInfo() output after the failure:

sessionInfo()
R version 3.5.2 (2018-12-20)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Red Hat Enterprise Linux Workstation release 6.10 (Santiago)

Matrix products: default
BLAS: .../binaries/R-3.5.2/lib64/R/lib/libRblas.so
LAPACK: .../binaries/R-3.5.2/lib64/R/lib/libRlapack.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] digest_0.6.18  dplyr_0.8.0.1  RxODE_0.8.0-9  testthat_2.0.1

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.0       crayon_1.3.4     withr_2.1.2      assertthat_0.2.0
 [5] R6_2.4.0         magrittr_1.5     pillar_1.3.1     cli_1.0.1       
 [9] rlang_0.3.1      lazyeval_0.2.1   rex_1.1.2        mvnfast_0.2.5   
[13] dparser_0.1.8    PreciseSums_0.3  glue_1.3.0       purrr_0.3.1     
[17] compiler_3.5.2   pkgconfig_2.0.2  memoise_1.1.0    tidyselect_0.2.5
[21] tibble_2.0.1    
>

**Comment by mattfidler** [comment]
Can you also post the version of Armadillo?
```R
library(RcppArmadillo)
packageVersion("RcppArmadillo")
```
It wasn't included in the sessionInfo()

**Comment by bruc** [comment]
> library(RcppArmadillo)
> packageVersion("RcppArmadillo")
[1] ‘0.9.200.7.1’

**Comment by bruc** [comment]
It appears to be the compiler -- I just rebuilt R and RxODE using gfortran 8.2.1 20180905 and there's no error with RxODE. I still have to upgrade back to 8.3.1 and repeat the test, but since Debian devel is probably running the most recent Gnu compiler, this seems a likely explanation.

**Comment by mattfidler** [comment]
Sigh.  These things are frustrating.  Thank you for providing an explanation @bruc

**Comment by bruc** [comment]
You're welcome.

Just for completeness: I have rebuilt R and RxODE with the GCC compiler suite (8.3.1 20190311) and it fails with the original error in DTRTI2.

**Comment by mattfidler** [comment]
Could you try the following for completeness:

1. Rebuild RxODE from github
2. Test to see if it works.  In theory if you installed the tests you could just do `rxTest("chol")` to reduce the number of tests.

I added some code to try to work around this issue, but I'm unsure if it works.

**Comment by bruc** [comment]
I just reloaded the 8.3.1 compiler suite, and built RxODE from github. The rxTest("chol") passes.

Just a suggestion: I wouldn't work around this problem -- your test code should confirm that the compiler is working correctly, and the previous code does that in a limited way. Perhaps the "configure" script should check for gfortran 8.3.1 and stop with an error if it finds it.

BTW, the DESCRIPTION file is missing the Author field, so "R CMD check" fails. However, I added the field to my local copy of the package, and then "R CMD check" is successful.

**Comment by mattfidler** [comment]
Thanks @bruc; Its good to know it will pass.

Unfortunately CRAN will delist you if you don't pass their tests.  Therefore, a work-around is OK.  However in this case it is a compiler bug, so they may be a bit more lenient.

I'm unsure why you don't see an Author field in the description.  I see it here:

https://github.com/nlmixrdevelopment/RxODE/blob/f95e937a87b6d1d71e550aeb2f001750cb4bdb36/DESCRIPTION#L4

I still need to make a reprex for the gfortran development team.

**Comment by mattfidler** [comment]
Hi @bruc 

If you would be so kind could you tell me what version of redhat you are using;  I will try to reproduce with a virtual machine.  I'm a redhat newbie, so the `yum` commands to install what I need to would also be helpful.

I believe I have created an armadillo (modified by dirk) example that would cause the failure, but I need to test:

```c++
// -*- mode: C++; c-indent-level: 4; c-basic-offset: 4; compile-command: "g++ -s -Wall -O3 -o arma_issue254 arma_issue254.cpp -larmadillo"
#include <iostream>
#include <armadillo>

using namespace arma;

int main(int argc, const char **argv){
  arma::Mat<double> A = arma::randu(4,4);
  A = A*A.t();
  A = chol(A);
  A = inv(trimatu(A)); // This is the line that gives the error
  A.print("A at end");
}
```

Otherwise, if you are willing you can test yourself;

```sh
sudo yum armadillo-devel
g++ -s -Wall -O3 -o arma_issue254 arma_issue254.cpp -larmadillo
./arma_issue254
```

If successfull it should output:
```sh
A at end
   1.0482  -0.7919  -1.9054  -0.8589
        0   1.5650  -2.3196  -0.2896
        0        0   3.5934  -2.8233
        0        0        0   2.6929
```

**Comment by bruc** [comment]
I'm not able to work on this again until Friday or the weekend.

I'm running the latest release of Redhat 6, but the key element here is the Developers Toolset 8, which is another subscription to  Redhat. It may also be available on Centos -- they had some earlier versions of these toolsets. 'yum install devtoolset-8\*' should do it. My package list is just under 3000 RPM's, but I doubt you need them all.

The problem with Author: was at the end of the DESCRIPTION file. The first author list was present in the copy that failed the "R CMD check", however the second list was missing. It was present in the CRAN version of RxODE.

**Comment by mattfidler** [comment]
These are my notes for what I'm trying on CentOS:
1. First use the osboxes centos 7 image
2. Do a sudo reposync
3. Install the [kernel libraries](https://linuxconfig.org/how-to-install-virtualbox-guest-additions-on-centos-7-linux) for the guest additions add-on
```sh
sudo rpm -Uvh https://dl.fedoraproject.org/pub/epel/epel-release-latest-7.noarch.rpm
sudo yum install perl gcc dkms kernel-devel kernel-headers make bzip2
```
4. "Insert" the guest addition addons
5. Install [gcc8](https://unix.stackexchange.com/questions/477360/centos-7-gcc-8-installation) with the following commands:

```sh
yum install centos-release-scl
yum install devtoolset-8-*
scl enable devtoolset-8 -- bash
```
6. Install the armadillo-devel
```sh
yum install armadillo-devel
```
7. As a normal user do:
```sh
scl enable devtoolset-8 -- bash
g++ -s -Wall -O3 -o arma_issue254 arma_issue254.cpp -larmadillo
```

This gives the normal results

```sh
[osboxes@osboxes tmp]$ ./arma_issue254 
A at end
   1.0482  -0.7919  -1.9054  -0.8589
        0   1.5650  -2.3196  -0.2896
        0        0   3.5934  -2.8233
        0        0        0   2.6929
```

I'm afraid I can't reproduce the issue on CentOS...

```

**Comment by mattfidler** [comment]
Note the g++ version:
```sh
[osboxes@osboxes tmp]$ g++ --version
g++ (GCC) 8.2.1 20180905 (Red Hat 8.2.1-3)
Copyright (C) 2018 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

**Comment by mattfidler** [comment]
Unfortunately, I am also unsure if the armadillo package was built with gcc 8.2.1. 

Do you know any way to check?

**Comment by mattfidler** [comment]
Nevermind I figured it out.  It isn't complied with gcc 8.2.1

**Comment by mattfidler** [comment]
Compile

1. Install git
```sh
sudo yum install git
```
2. download lapack and compile
```
git clone https://github.com/Reference-LAPACK/lapack-release.git
cd lapack-release
cp make.inc.example make.inc
make
```

After a bit I get:
```sh
NEP: Testing Nonsymmetric Eigenvalue Problem routines
./EIG/xeigtstz < nep.in > znep.out 2>&1
/bin/sh: line 1: 26579 Segmentation fault      (core dumped) ./EIG/xeigtstz < nep.in > znep.out 2>&1
make[1]: *** [Makefile:464: znep.out] Error 139
make[1]: Leaving directory '/home/osboxes/tmp/lapack-release/TESTING'
make: *** [Makefile:43: lapack_testing] Error 2
[osboxes@osboxes lapack-release]$ 
```

**Comment by bruc** [comment]
gcc 8.2.1 works -- the problem came with 8.3.1.

**Comment by mattfidler** [comment]
Ah.  Well CentOS only has 8.2.1

**Comment by bruc** [comment]
I tried running your program, I get a different matrix than you did, but the result is unchanged between the 8.2.1 and 8.3.1 compilers. Here's the specifics:
```
bruc@liberty:~/projects/RxODE$ gfortran -v
Using built-in specs.
COLLECT_GCC=gfortran
COLLECT_LTO_WRAPPER=/opt/rh/devtoolset-8/root/usr/libexec/gcc/x86_64-redhat-linux/8/lto-wrapper
Target: x86_64-redhat-linux
Configured with: ../configure --enable-bootstrap --enable-languages=c,c++,fortran,lto --prefix=/opt/rh/devtoolset-8/root/usr --mandir=/opt/rh/devtoolset-8/root/usr/share/man --infodir=/opt/rh/devtoolset-8/root/usr/share/info --with-bugurl=http://bugzilla.redhat.com/bugzilla --enable-shared --enable-threads=posix --enable-checking=release --enable-multilib --with-system-zlib --enable-__cxa_atexit --disable-libunwind-exceptions --enable-gnu-unique-object --enable-linker-build-id --with-gcc-major-version-only --with-linker-hash-style=gnu --with-default-libstdcxx-abi=gcc4-compatible --enable-plugin --enable-initfini-array --with-isl=/builddir/build/BUILD/gcc-8.3.1-20190311/obj-x86_64-redhat-linux/isl-install --disable-libmpx --with-mpc=/builddir/build/BUILD/gcc-8.3.1-20190311/obj-x86_64-redhat-linux/mpc-install --with-tune=generic --with-arch_32=i686 --build=x86_64-redhat-linux
Thread model: posix
gcc version 8.3.1 20190311 (Red Hat 8.3.1-3) (GCC) 

bruc@liberty:~/projects/RxODE$ cat arma_issue254.cpp
// -*- mode: C++; c-indent-level: 4; c-basic-offset: 4; compile-command: "g++ -s -Wall -O3 -o arma_issue254 arma_issue254.cpp -larmadillo"
#include <iostream>
#include <armadillo>

using namespace arma;

int main(int argc, const char **argv){
  arma::Mat<double> A = arma::randu(4,4);
  A = A*A.t();
  A = chol(A);
  A = inv(trimatu(A)); // This is the line that gives the error
  A.print("A at end");
}
bruc@liberty:~/projects/RxODE$ g++ -s -Wall -O3 -o arma_issue254 arma_issue254.cpp -larmadillo
bruc@liberty:~/projects/RxODE$ ./arma_issue254
A at end
   0.7565  -0.8260  -0.5974  -2.6828
        0   1.6928  -3.1111  -3.4607
        0        0   2.5606  -2.0301
        0        0        0   5.6213
bruc@liberty:~/projects/RxODE$ 

```

I got the same results with 8.2.1 compilers.

**Comment by mattfidler** [comment]
The difference in matrix values is expected.  (Although Dirk and I had the same matrices, I didn't seed the random number generator, but apparently it could be platform specific). 

It seems odd that the 8.3.1 compiler works with this example.  The code that is breaking is the inversion of upper/lower triangular matrices.  

Since I have a "fix" of some sort, for now I will close this, but keep my eyes open for anything new.

**Comment by bruc** [comment]
I have opened a case with Redhat to look at this problem. It may be a while before they resolve it. I'll let you know what happens.

**Comment by conradsnicta** [comment]
Fixed in Armadillo 9.500-RC1: http://arma.sourceforge.net/download.html

**Comment by mattfidler** [comment]
Thanks @conradsnicta !!!
